proposals/nnnn-implicit-open-existentials.md

```diff
@@ -0,0 +1,387 @@
+
+This section describes the details of opening an existential and then type-erasing back to an existential. Most of these details of this change should be invisible to the user, and manifest only as the ability to use existentials with generics in places where the code would currently be rejected.
+
+### When can we open an existential?
+
+To open an existential, the argument (or source) must be of existential type (e.g., `any P`) or existential metatype (e.g., `any P.Type`) and must be provided to a parameter (or target) whose type involves a generic parameter that can bind directly to the underlying type of the existential. This means that, for example, we can open an existential when its underlying type would directly bind to a generic parameter:
+
+```swift
+protocol P {
+  associatedtype A
+  
+  func getA() -> A
+}
+
+func openSimple<T: P>(_ value: T) { }
+
+func testOpenSimple(p: any P) {
+  openSimple(p) // okay, opens 'p' and binds 'T' to its underlying type
+}
+```
+
+It's also possible to open an `inout` parameter. The generic function will operate on the underlying type, and can (e.g.) call `mutating` methods on it, but cannot change its *dynamic* type because it doesn't have access to the existential box:
+
+```swift
+func openInOut<T: P>?(_ value: inout T) { }
+func testOpenInOut(p: any P) {
+  var mutableP: any P = p
+  openInOut(&mutableP) // okay, opens to 'mutableP' and binds 'T' to its underlying type
+}
+```
+
+However, we cannot open when there might be more than one value of existential type or no values at all, because we need to be guaranteed to have a single underlying type to infer. Here are several such examples where the generic parameter is in a structural position:
+
+```swift
+func cannotOpen1<T: P>(_ array: [T]) { .. }
+func cannotOpen2<T: P>(_ a: T, _ b: T) { ... }
+func cannotOpen3<T: P>(_ values: T...) { ... }
+
+struct X<T> { }
+func cannotOpen4<T: P>(_ x: X<T>) { }
+
+func cannotOpen5<T: P>(_ x: T, _ a: T.A) { }
+
+func cannotOpen6<T: P>(_ x: T?) { }
+
+func testCannotOpenMultiple(array: [any P], p1: any P, p2: any P, xp: X<any P>, pOpt: (any P)?) {
+  cannotOpen1(array)         // each element in the array can have a different underlying type, so we cannot open
+  cannotOpen2(p1, p2)        // p1 and p2 can have different underlying types, so there is no consistent binding for 'T'
+  cannotOpen3(p1, p2)        // similar to the case above, p1 and p2 have different types, so we cannot open them
+  cannotOpen4(xp)            // cannot open the existential in 'X<any P>' there isn't a specific value there.
+  cannotOpen5(p1, p2.getA()) // cannot open either argument because 'T' is used in both parameters
+  cannotOpen6(pOpt)         // cannot open the existential in '(any P)?' because it might be nil, so there would not be an underlying type
```

Comment:
```suggestion
  cannotOpen6(pOpt)          // cannot open the existential in '(any P)?' because it might be nil, so there would not be an underlying type
```